rfcs/0042-local-test-framework.md

```diff
@@ -0,0 +1,272 @@
+# Summary:
+
+Currently, the test integration framework has the capabilities to upload and test a network to a cloud environment (specifically GCP) but in addition to this, we want to add functionality to deploy and test a network on a local user machine. The cloud integration uses Terraform and Helm to deploy to a GKE environment to deploy the network and specified nodes. While this method works for cloud environments, we would like a more lightweight solution to run locally. Thus, we have chosen `Docker Swarm` as the tool of choice for container orchestration. 
+
+Docker swarm can configure a "swarm" on a local machine to deploy and manage containers on that specific swarm. Docker Swarm takes as input a `docker-compose` file in which all container information is specified and will handle the deployment of all containers on that local swarm. When we want to run a network test locally, we can create a swarm and have all the containers deploy via a `docker-compose.json` file that is built from specified network configurations. Docker swarm also gives us the ability to get logs of all the containers running in an aggregated way, meaning we do not have to query individual containers for their logs. This gives us a way to apply event filters to specific nodes (block producers, snark workers, seed nodes, etc), and check for test success/failure in a portable way.
+
+# Requirements
+
+The new local testing framework should be run on a user's local system using Docker as its main engine to create a network and spawn nodes. This feature will be built on top of the existing Test Executive which runs our cloud integration tests. By implementing the interface specified in `src/lib/integration_test_lib/intf.ml`, we will have an abstract way to specify different testing engines when running the Test Executive. 
+
+The specific interface to implement would be:
+
+```ocaml
+(** The signature of integration test engines. An integration test engine
+   *  provides the core functionality for deploying, monitoring, and
+   *  interacting with networks.
+   *)
+  module type S = sig
+    (* unique name identifying the engine (used in test executive cli) *)
+    val name : string
+
+    module Network_config : Network_config_intf
+
+    module Network : Network_intf
+
+    module Network_manager :
+      Network_manager_intf
+      with module Network_config := Network_config
+       and module Network := Network
+
+    module Log_engine : Log_engine_intf with module Network := Network
+  end
+end
+```
+
+To implement this interface, a new subdirectory will be created in `src/lib` named `integration_local_engine` to hold all implementation details for the local engine.
+
+The new local testing engine must implement all existing features which include:
+
+- Starting/Stopping nodes dynamically
```

Comment:
Just wanted to mention as a note here: If you continue using the puppeteer container (which I recommend), then this is very easy to do. You can peek at how this works in the cloud engine, but starting/stopping a puppeteered node is as simple as running a shell script in the container which communicates with the puppeteer process.